src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController_ModelComputation.cs

```diff
@@ -99,70 +116,67 @@ private async Task<NavigationBarModel> ComputeModelAsync(ITextSnapshot snapshot,
                         items.Do(i => i.InitializeTrackingSpans(snapshot));
                         var version = await document.Project.GetDependentSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
 
-                        _lastCompletedModel = new NavigationBarModel(items, version, languageService);
-                        return _lastCompletedModel;
+                        return new NavigationBarModel(items.ToImmutableArray(), version, languageService);
                     }
                 }
             }
 
-            _lastCompletedModel ??= new NavigationBarModel(SpecializedCollections.EmptyList<NavigationBarItem>(), new VersionStamp(), null);
-            return _lastCompletedModel;
+            return new NavigationBarModel(ImmutableArray<NavigationBarItem>.Empty, new VersionStamp(), null);
         }
 
-        private Task<NavigationBarSelectedTypeAndMember> _selectedItemInfoTask;
-        private CancellationTokenSource _selectedItemInfoTaskCancellationSource = new();
-
         /// <summary>
         /// Starts a new task to compute what item should be selected.
         /// </summary>
-        private void StartSelectedItemUpdateTask(int delay, bool updateUIWhenDone)
+        private void StartSelectedItemUpdateTask(int delay)
         {
             AssertIsForeground();
 
             var currentView = _presenter.TryGetCurrentView();
-            if (currentView == null)
-            {
+            var subjectBufferCaretPosition = currentView?.GetCaretPoint(_subjectBuffer);
+            if (!subjectBufferCaretPosition.HasValue)
                 return;
-            }
 
             // Cancel off any existing work
             _selectedItemInfoTaskCancellationSource.Cancel();
             _selectedItemInfoTaskCancellationSource = new CancellationTokenSource();
-
             var cancellationToken = _selectedItemInfoTaskCancellationSource.Token;
-            var subjectBufferCaretPosition = currentView.GetCaretPoint(_subjectBuffer);
 
-            if (!subjectBufferCaretPosition.HasValue)
-            {
+            var asyncToken = _asyncListener.BeginAsyncOperation(GetType().Name + ".StartSelectedItemUpdateTask");
+            var selectedItemInfoTask = DetermineSelectedItemInfoAsync(_modelTask, delay, subjectBufferCaretPosition.Value, cancellationToken);
+            selectedItemInfoTask.CompletesAsyncOperation(asyncToken);
+        }
+
+        private async Task DetermineSelectedItemInfoAsync(
+            Task<NavigationBarModel> lastModelTask,
+            int delay,
+            SnapshotPoint caretPosition,
+            CancellationToken cancellationToken)
+        {
+            // First wait the delay before doing any other work.  That way if we get canceled due to other events (like
+            // the user moving around), we don't end up doing anything, and the next task can take over.
```

Comment:
as mentioned in the above comment, this is kind of what I was expecting for ComputeModelAfterDelayAsync so that we don't just keep queueing up delays

Reply:
the issue with ComputeModelAfterDelay is that we do have to pass along the model we have computed from before.  This is needed so that the work to compute the model actually can check if it can bail out and not do anything.